File: python/questions/BSTIterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
     
    def __init__(self, root: Optional[TreeNode]):
        self.ins = []
        self.pushLeft(root)
        
    def next(self) -> int:
        curn = self.ins[-1]
        self.ins.pop()
        self.pushLeft(curn.right)
        return curn.val

    def hasNext(self) -> bool:
        return len(self.ins) > 0

    def pushLeft(self, node: Optional[TreeNode]):
        if node is None:
            return
        self.ins.append(node)
        self.pushLeft(node.left)
# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: typescript/questions/BSTIterator.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class BSTIterator {
    private ins: TreeNode[];
    
    private pushLeft(node: TreeNode | null) {
	while (node != null) {
	    this.ins.push(node);
	    node = node.left;
	}
    }
    
    constructor(root: TreeNode | null) {
        this.ins = [];
	this.pushLeft(root);
    }

    next(): number {
        const curn = this.ins.pop();
	this.pushLeft(curn.right);
	return curn.val;
    }

    hasNext(): boolean {
        return this.ins.length > 0;
    }

}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
